### PR TITLE
improve eliding in the flamegraph

### PR DESCRIPTION
**Pull request by lievenhey**
in some cases the flamegraph contains only std::... and not the function
name.
This patch allows the user to change the eliding to ElideLeft so that
you can see the function name

### CODE REVIEW

File: src/settings.cpp [hunk 1]
```diff
@@ -218,7 +218,39 @@ void Settings::loadFromFile()
...
+
+    connect(this, &Settings::flamegraphTextElideModeChanged, [sharedConfig](const Qt::TextElideMode& mode) {
+        sharedConfig->group("Flamegraph").writeEntry("elideMode", elideLookup[mode]);
+    });
+}
+
+void Settings::setFlamegraphTextElideMode(const Qt::TextElideMode& mode)
+{
+    // we only support ElideLeft and ElideRight
+    Q_ASSERT(mode == Qt::ElideLeft || Qt::ElideRight);
```
Comment (milianw) [inline src/settings.cpp:250]: this check is always true because Qt::ElideRight is always true

**Review by milianw — CHANGES_REQUESTED** [review]
instead of making this configurable, I'd say we should make the elision more useful?

 Neither left nor right elision is ideal for absurd symbols like

```
asdf_namespace::foobar<asdf, yxcvyxcv>::blablub(someotherreallylongnames, ...)
```

ideally, we would detect such long names and then iteratively elide them, such that:

a) first we drop the function arguments and the above gets turned into
```
asdf_namespace::foobar<asdf, yxcvyxcv>::blablub(...)
```
b) if that's still too long, we left remove template arguments (we have that already, reuse that logic)
```
asdf_namespace::foobar<...>::blablub(...)
```
c) finally if that's still too long, we left-elide the remainder to produce e.g. (depending on width):
```
...::blablub(...)
```

ideally we also don't use three dots but the utf8 sequence for the ellision, i.e. `…`

File: tests/modeltests/tst_models.cpp [hunk 2]
```diff
@@ -651,6 +651,33 @@ private slots:
...
+        QTest::addRow("elide templates") << 700 << "asdf_namespace::foobar<…>::blablub(…) const";
+        QTest::addRow("elide symbol") << 350 << "…obar<…>::blablub(…) const";
+    }
+
+    void testSymbolEliding()
+    {
+        const QFontMetrics metrics(QFont(QStringLiteral("monospace"), 10));
+
+        const QString testSymbol =
+            QLatin1String("asdf_namespace::foobar<asdf, yxcvyxcv>::blablub(someotherreallylongnames) const");
```
Comment (milianw) [inline tests/modeltests/tst_models.cpp:674]: QStringLiteral

File: src/util.cpp [hunk 3]
```diff
@@ -32,6 +33,28 @@
 
 #include <KParts/ReadOnlyPart>
 
+namespace {
+QString elideArguments(const QString symbolText)
```
Comment (milianw) [inline src/util.cpp:37]: const&

File: src/util.cpp [hunk 4]
```diff
@@ -32,6 +33,28 @@
...
+{
+    int startEliding = symbolText.indexOf(QLatin1Char('('));
```
[the remark this answers is not in the document]
Comment (milianw) [reply src/util.cpp:39]: https://invent.kde.org/kdevelop/kdevelop/-/merge_requests/378 contains a bunch of other fancy identifiers that are hard to parse...

[the remark this answers is not in the document]
Comment (lievenhey) [reply src/util.cpp]: I added some more logic, so that hotspot is now able to elide those monstrosities

File: src/util.cpp [hunk 5]
```diff
@@ -32,6 +33,66 @@
 
 #include <KParts/ReadOnlyPart>
 
+namespace {
+struct ElidedText
+{
+    QString first;
```
Comment (milianw) [inline src/util.cpp:39]: `std::pair<QString, QString>`? or use better names

File: src/util.cpp [hunk 6]
```diff
@@ -32,6 +33,28 @@
...
+
+    int endEliding = symbolText.indexOf(QLatin1Char(')'), startEliding);
+
+    if (endEliding - startEliding == 1) {
+        // don't elide if we have () in symbol
+        return symbolText;
+    }
+
+    // QLatin1String does not work with …
+    const auto elideSymbol = QString::fromUtf8("…");
```
[the remark this answers is not in the document]
Comment (milianw) [reply src/util.cpp:91]: this is not resolved

File: src/util.cpp [hunk 7]
```diff
@@ -57,6 +57,9 @@ QString collapseTemplate(const QString& str, int level)
         return false;
     };
 
+    // QLatin1String does not work with …
+    const auto elideString = QString::fromUtf8("<…");
```
[the remark this answers is not in the document]
Comment (milianw) [reply src/util.cpp:122]: not resolved

File: tests/modeltests/tst_models.cpp (hunk 2 above)
Comment (milianw) [reply tests/modeltests/tst_models.cpp:673]: not done

File: tests/modeltests/tst_models.cpp [hunk 8]
```diff
@@ -651,6 +651,41 @@ private slots:
...
+        QFETCH(QString, elidedSymbol);
+
+        QCOMPARE(Util::elideSymbol(testSymbol, metrics, maxWidth), elidedSymbol);
+    }
+
+    void testSymbolElidingParanthese()
+    {
+        const QFontMetrics metrics(QFont(QStringLiteral("monospace"), 10));
+
+        const QString symbol = QLatin1String("Foo<&bar::operator()>::asdf<XYZ>(blabla<&foo::operator(), ')', '('>)");
```
Comment (milianw) [inline tests/modeltests/tst_models.cpp:685]: dito

File: src/util.cpp [hunk 9]
```diff
@@ -32,6 +33,66 @@
...
+ElidedText elideArguments(const QString symbolText)
+{
+    QChar quoteChar = QLatin1Char('"');
+    int numQuotes = 0;
+    int depth = 0;
+
+    int endEliding = symbolText.length();
+    int startEliding = 0;
+
+    for (int i = symbolText.length() - 1; i > 0; i--) {
```
Comment (milianw) [inline src/util.cpp:52]: please add some basic comments that document what this code is doing on a high level

**Review by milianw — CHANGES_REQUESTED** [review]

**Review by milianw — APPROVED** [review]
lgtm, thanks
